test/Test_Elaboration.re

```diff
@@ -7,90 +7,105 @@ let dhexp_typ = testable(Fmt.using(Exp.show, Fmt.string), DHExp.fast_equal);
 
 let ids = List.init(12, _ => Id.mk());
 let id_at = x => x |> List.nth(ids);
+
 let mk_map = Statics.mk(CoreSettings.on, Builtins.ctx_init);
 let dhexp_of_uexp = u => Elaborator.elaborate(mk_map(u), u) |> fst;
 let alco_check = dhexp_typ |> Alcotest.check;
 
-let u1: Exp.t = {ids: [id_at(0)], term: Int(8), copied: false};
-let single_integer = () =>
-  alco_check("Integer literal 8", u1, dhexp_of_uexp(u1));
-
-let u2: Exp.t = {ids: [id_at(0)], term: EmptyHole, copied: false};
-let empty_hole = () => alco_check("Empty hole", u2, dhexp_of_uexp(u2));
-
-let u3: Exp.t = {
-  ids: [id_at(0)],
-  term: Parens({ids: [id_at(1)], term: Var("y"), copied: false}),
-  copied: false,
-};
-
-let free_var = () => alco_check("free variable", u3, dhexp_of_uexp(u3));
+let alco_check_menhir = (name: string, dhexp: string, uexp: Term.Exp.t) =>
+  alco_check(
+    name,
+    Haz3lmenhir.Conversion.Exp.of_menhir_ast(
+      Haz3lmenhir.Interface.parse_program(dhexp),
+    ),
+    dhexp_of_uexp(uexp),
+  );
 
-let u4: Exp.t =
+//Test for a let function
+let let_fun_uexp: Exp.t =
   Let(
-    Tuple([Var("a") |> Pat.fresh, Var("b") |> Pat.fresh]) |> Pat.fresh,
-    Tuple([Int(4) |> Exp.fresh, Int(6) |> Exp.fresh]) |> Exp.fresh,
-    BinOp(Int(Minus), Var("a") |> Exp.fresh, Var("b") |> Exp.fresh)
-    |> Exp.fresh,
-  )
-  |> Exp.fresh;
-
-let let_exp = () =>
-  alco_check("Let expression for tuple (a, b)", u4, dhexp_of_uexp(u4));
-
-let u5 =
-  BinOp(Int(Plus), Bool(false) |> Exp.fresh, Var("y") |> Exp.fresh)
-  |> Exp.fresh;
-
-let d5 =
-  BinOp(
-    Int(Plus),
-    FailedCast(Bool(false) |> Exp.fresh, Bool |> Typ.fresh, Int |> Typ.fresh)
-    |> Exp.fresh,
     Cast(
-      Var("y") |> Exp.fresh,
+      Var("f") |> Pat.fresh,
+      Arrow(Int |> Typ.fresh, Int |> Typ.fresh) |> Typ.fresh,
       Unknown(Internal) |> Typ.fresh,
-      Int |> Typ.fresh,
+    )
+    |> Pat.fresh,
+    Fun(
+      Var("x") |> Pat.fresh,
+      BinOp(Int(Plus), Int(1) |> Exp.fresh, Var("x") |> Exp.fresh)
+      |> Exp.fresh,
+      None,
+      None,
     )
     |> Exp.fresh,
+    Int(55) |> Exp.fresh,
   )
   |> Exp.fresh;
 
-let bin_op = () =>
-  alco_check(
-    "Inconsistent binary integer operation (plus)",
-    d5,
-    dhexp_of_uexp(u5),
+let let_fun_str = "
+let f =
+    fun x ->
+        1 + x
+    f
```

Comment:
I'm a bit confused syntactically why `f` is below the function. Is this to capture the name? And if so do we intend on this being incompatible with the regular hazel parser

Reply:
That's to cover the `option(Var.t)` at the end of the `Fun` exp. I'm not too sure what that option represents/how it gets inserted in elaboration so I allowed it to be optionally specified textually for the menhir parser. Do you happen to know what it represents/how I should do this properly? Thanks!

Reply:
I think that's the name of the function that gets pulled from the let binding. Let me confirm that and see where it's happening. Unfortunately I don't think there's anywhere in the existing syntax that differentiates these cases.

Reply:
Okay, so the function name gets added during elaboration https://github.com/hazelgrove/hazel/blob/dev/src/haz3lcore/dynamics/Elaborator.re#L292. So there's no place in the syntax text that represents this currently.

I'm not sure how we want to represent this in the Menhir parser. We can just remove it and have this be compatible with `MakeTerm` but then we couldn't use the parser for the expectation side of the elaboration tests (without changing the assertion equality to ignore function names). 

For now I'm okay with either ignoring the names in the parser or having some artificial syntactic representation of a named function.  I'm not sure if just putting the name after the function will be confusing or not.

@cyrus- or @Negabinary  do you have any other thoughts?

Reply:
https://github.com/hazelgrove/hazel/pull/1409/files#diff-6967fe38adf76426a01468c788252c2355461e812298332ce6e3e282021ee705R46 might also be useful for any confusion on how the existing parser is handling cases

Reply:
having an artificial syntax for the function name seems fine to me

Reply:
@green726 feel free to come up with your own representation, but off my head I like `fun x -> 1 + x` for anonymous and `named_fun f x -> 1+x` for named.

Reply:
Got it - your representation sounds good to me. I'll make that change.